components/modules/file.c

```diff
@@ -383,32 +383,27 @@ static int file_stat( lua_State* L )
 // g_read()
 static int file_g_read( lua_State* L, int n, int16_t end_char, int fd )
 {
-  static char *heap_mem = NULL;
-  // free leftover memory
-  if (heap_mem) {
```

Comment:
This code was added to free memory which was left allocated by a previous call. Standard execution flow doesn't need that, but `heap_mem` isn't freed in case `lua_pushlstring()` throws a memory exception as noted in https://github.com/nodemcu/nodemcu-firmware/pull/1541#issuecomment-254654541.

If we keep considering this corner case then `bufsize` would also need to be declared `static`.

Reply:
@devsaurus I added a new function to make a protected call to `lua_pushlstring()` and avoid this issue. The new `luaX_pushlstring()` returns nonzero instead of throwing an exception, allowing you to free memory. This function could be useful elsewhere.

Please check it out and let me know :-)

Reply:
Excellent solution!